drivers/adc/ad713x/ad713x.c

```diff
@@ -258,137 +300,278 @@ int32_t ad713x_mag_phase_clk_delay(ad713x_dev *dev,
+	}
+	if (dev->gpio_pnd) {
+		ret = gpio_remove(dev->gpio_pnd);
+		if(IS_ERR_VALUE(ret))
+			return FAILURE;
+	}
+	if (dev->gpio_resetn) {
+		ret = gpio_remove(dev->gpio_resetn);
+		if(IS_ERR_VALUE(ret))
+			return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the wideband filter bandwidth for every channel.
+ *        ad713x_init() helper function.
+ * @param [in] dev - AD713X device handler.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+static int32_t ad713x_init_chan_bw(struct ad713x_dev *dev)
+{
+	int8_t i;
+	int32_t ret;
+
+	for (i = CH3; i >= 0; i--) {
+		ret = ad713x_wideband_bw_sel(dev, i, 0);
+		if (IS_ERR_VALUE(ret))
+			return FAILURE;
+	}
+
+	return SUCCESS;
 }
 
 /**
- * Initialize the device.
+ * @brief Initialize the device.
  * @param device - The device structure.
  * @param init_param - The structure that contains the device initial
  *                     parameters.
- * @return 0 in case of success, negative error code otherwise.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
  */
-int32_t ad713x_init(ad713x_dev **device,
-		    ad713x_init_param init_param)
+int32_t ad713x_init(struct ad713x_dev **device,
+		    struct ad713x_init_param *init_param)
 {
-	ad713x_dev *dev;
-	uint8_t buf[3];
+	struct ad713x_dev *dev;
 	int32_t ret;
+	uint8_t data;
 
-	dev = (ad713x_dev *)malloc(sizeof(*dev));
+	dev = (struct ad713x_dev *)calloc(1, sizeof(*dev));
 	if (!dev)
-		return -1;
-
-	/* SPI */
-	ret = spi_init(&dev->spi_desc, &init_param.spi_init);
-	/* GPIO */
-	ret |= gpio_get(&dev->gpio_mode1, init_param.gpio_mode1);
-	ret |= gpio_get(&dev->gpio_mode2, init_param.gpio_mode2);
-	ret |= gpio_get(&dev->gpio_dclkmode, init_param.gpio_dclkmode);
-	ret |= gpio_get(&dev->gpio_dclkio1, init_param.gpio_dclkio1);
-	ret |= gpio_get(&dev->gpio_dclkio2, init_param.gpio_dclkio2);
-	ret |= gpio_get(&dev->gpio_resetn1, init_param.gpio_resetn1);
-	ret |= gpio_get(&dev->gpio_resetn2, init_param.gpio_resetn2);
-	ret |= gpio_get(&dev->gpio_pnd1, init_param.gpio_pnd1);
-	ret |= gpio_get(&dev->gpio_pnd2, init_param.gpio_pnd2);
+		return FAILURE;
+
+	if (!init_param->spi_common_dev) {
+		ret = spi_init(&dev->spi_desc, &init_param->spi_init_prm);
+		if (IS_ERR_VALUE(ret))
+			goto error_dev;
+	} else {
+		dev->spi_desc = calloc(1, sizeof *dev->spi_desc);
+		dev->spi_desc->chip_select = init_param->spi_init_prm.chip_select;
+		dev->spi_desc->extra = init_param->spi_common_dev->extra;
+		dev->spi_desc->max_speed_hz = init_param->spi_init_prm.max_speed_hz;
+		dev->spi_desc->mode = init_param->spi_init_prm.mode;
+	}
+
+	ret = ad713x_init_gpio(dev, init_param);
+	if(IS_ERR_VALUE(ret))
+		goto error_gpio;
 	/*
 	 * Tie this pin to IOVDD for master mode operation,
 	 * tie this pin to IOGND for slave mode operation.
 	 */
-	gpio_set_value(dev->gpio_mode1, GPIO_HIGH);
-	gpio_set_value(dev->gpio_mode1, GPIO_LOW);
+	gpio_direction_output(dev->gpio_mode, init_param->mode_master_nslave);
```

Comment:
Why don't you put this also in gpio_init?

Reply:
Good point. I will make this change.